doc/designs/application-specific-passwords.md

```diff
@@ -0,0 +1,243 @@
+* the user will set application-specific passwords as usual (having to fulfil
+  password policies in place), and/or
+* they will be generated by FreeIPA (possibly only for one use, being an OTP).
+  It is usually done so by web services using application-specific passwords as
+  a fallback mechanism when 2FA is activated, but the client application does
+  not support it.
+
+### Permissions management for application-specific passwords
+
+The user will be able to configure permissions based on the password
+used to log into the account:
+* logging in using the primary password (or the master 2FA token)
+would grant full access to the account, including configuring all
+the other passwords as well as access to all services for which
+that account is used,
+* logging in with an application-specific password would allow
+access only to a specified subset of systems and services that
+password is allowed access to (for example, it could grant access
+only to log in to the user's mail server).
+
+By default, the password will have no permissions assigned, being inactive in
+practice, with the user then explicitly assigning permissions for it at his/her
+discretion. This may range from access to only one particular service up to
+either (TBD):
+* access rights on par with his primary password, or
+* access to all systems and services, but not to the account management.
+
+It would be better for this to be at least to some extent separate from the
+general access control, as permissions for passwords will be set by the user,
+while permissions for the whole account are often set by other people. As a part
+of this, it may be desirable to implement an attribute determining whether
+default and managed permissions apply for the given user's password, or not.
+
+There is another option: in theory, authentication indicators could be used for
+access management for multiple passwords for a single user - it would then be
+best to differentiate 'levels' of passwords. This might (depending on
+implementation) facilitate less fine-grained access control for varius
+passwords, but would probably be easier to implement.
+
+If needed, self-service ACIs and/or delegation ACIs will be extended as well
+to support application-specific passwords.
+
+#### HBAC for application-specific passwords
+
+There are two options how to make HBAC distinguish between various passwords:
+* Assign permissions for access to hosts as a part of the permissions management
+  for the user's passwords. HBAC would be extended to check these permissions.
+* Extend HBAC rules to *(host, service, user, password id/password level)*
+  so that they will directly accommodate this functionality. The attribute
+  identifying a user's password or a password level:
+  * can either be an optional field or it can have a special wildcard value denoting
+    'all user's passwords',
+  * will be administered by the user (in line with permissions management for
+    user's passwords in general).
```

Comment:
if we introduce the use of 2 component principals, and we make sssd understand them, then we could always assign a separate principal to each app, (user/appname) and then HBAC could use these principals for permissioning.
Need to decide whether users can invent whatever application name, or if pre-approved appnames need to be configured by admins.
Or perhaps even have app-namespaces where admins must configure "official appnames" but users can use a 'custom-' prefix to create additional app passwords for non official applications.
Admn could then set more restrictive permissions on those custom app passwords, and provide potential more relaxed permissions on "vetted" applications (assuming it makes sense to do so, it may not).

Reply:
We also need to investigate and change assumptions in IPA code. In general we assume that
* a two component principal with ``host`` as first component is always a host principal (``host/fqdn``)
* all other two component principal are always service principals (``service/fqdn``)
* single component principals are always users